### ISSUE TITLE
Enhancement: [no-unnecessary-condition] additionally check `typeof` expression when it possible

### ISSUE THREAD

**Issue by zanminkian** [typescript-eslint/typescript-eslint#11404]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unnecessary-condition/

### Description

Let's say:
```ts
function foo(): string | undefined {
  // bla bla bla
}
const bar = foo();
if (typeof bar === 'string') { // It's ok
  // bla bla bla
}
```

One day, I refactored the `foo` function to return string only:

```ts
function foo(): string {
  // bla bla bla
}
const bar = foo();
if (typeof bar === 'string') { // Smelly code. bar is always string. But no-unnecessary-condition does not report it
  // bla bla bla
}
```

### Fail

```TypeScript
const str:string = 'foo';
if (typeof str === 'string') {
  // bla bla bla
}
```

### Pass

```TypeScript
const str:string|undefined = 'foo';
if (typeof str === 'string') {
  // bla bla bla
}
```

**Comment by bradzacher** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/7064
